### DISCUSSION TITLE
How to get count of approved pull request using github actions

### DISCUSSION THREAD

**Discussion by so-amuk** [community/community#25374] [Actions]
<p>Hello everyone, is there any way of counting the approved PR using github actions?  I found this API<br>
<a href="https://api.github.com/repos/projectname/reponame/pulls/16000/reviews" rel="nofollow noopener">https://api.github.com/repos/projectname/reponame/pulls/16000/reviews</a> . But this give me the complete history of PR like reviewer when commented on it and when he approved it,  which is good. But how can  I count the total number of approvals of specific PR?<br>
Any suggestions please?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/so-amuk">@so-amuk</a>,<br>
Currently, from the API (<a href="https://developer.github.com/v3/pulls/review_requests/#list-review-requests" rel="nofollow noopener"><strong>List review requests</strong></a>) or the <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener"><strong>github context</strong></a>, we only can get the pending approvals (the reviewers who have not approved the PR).</p>
<p>So every time when you execute the API or view the github context of a latest workflow run for the PR, you may get different result with previous.</p>
<p>Here I have two workarounds that are relying on <a href="https://stedolan.github.io/jq/manual/" rel="nofollow noopener"><strong>jq</strong></a>:<br>
<strong>(1)</strong> Read from the github context like as below.</p>
<pre><code class="lang-yaml">echo '${{ toJson(github.event.pull_request.requested_reviewers) }}' | jq '. | length'
</code></pre>
<p>After you open a PR and have added all the reviewers on the PR, in the first workflow run for the PR, you can use this step to get the total count of the reviewers. Store and save this number via some way (such as <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows" rel="nofollow noopener"><strong>cache</strong></a>) so that you can use it in the subsequent workflow runs for the PR.</p>
<p>In the subsequent workflow runs, if some reviewers have approved the PR, you will get the count of the rest reviewers who have not approved the PR. Subtract this count from the saved total count to get the count of approved.</p>
<p><strong>(2)</strong> Read from the Response body of the API (<a href="https://developer.github.com/v3/pulls/reviews/#list-reviews-on-a-pull-request" rel="nofollow noopener"><strong>List reviews on a pull request</strong></a>).</p>
<pre><code class="lang-yaml">$response | ConvertTo-Json | jq '.[] | select(.state=="APPROVED") | length'
</code></pre>
<p>Every time when you execute this, it should return the count of approved.</p>

**Comment by so-amuk** [comment]
<p>Thanks for the answer. Can you please tell me how do you store count number in cache and use it?</p>
<p><a class="mention" href="/brightran">@brightran</a></p>

**Comment by so-amuk** [comment]
<p><a class="mention" href="/brightran">@brightran</a> I tried using 2nd approach here.  I get one issue here. If single user has approved the same PR 10 times, then I am getting 10 Approvals for same user in this case.<br>
The output would be something like:<br>
{<br>
“body”: “”,<br>
“state”: “APPROVED”,<br>
},<br>
{<br>
“body”: “”,<br>
“state”: “APPROVED”,<br>
},<br>
{<br>
“body”: “”,<br>
“state”: “APPROVED”,<br>
}<br>
It shows history of the PR.<br>
How to handle it?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/so-amuk">@so-amuk</a>,<br>
<strong>Workaround 1:</strong> (update)</p>
<pre><code class="lang-yaml">- name: get the count of requested reviewers
  run: |
    length=$(echo $requested_reviewers | jq '. | length')
    echo "length = $length"
  env:
    requested_reviewers: ${{ toJson(github.event.pull_request.requested_reviewers) }}
</code></pre>
<blockquote>
<p><strong>Can you please tell me how do you store count number in cache and use it?</strong></p>
</blockquote>
<p>You can simply write the total count of the reviewers into a text file, then save this text file as a cache with a specified key (for example, <strong>key: PR-&lt;PR_number&gt;-total-reviewers</strong>). In the subsequent workflow runs, you can restore this cache to the workspace and read the total count of the reviewers from the text file.<br>
However, if some reviewers are added or removed halfway, there may be problems with the calculated numbers in the subsequent workflow runs. So the premise of the 1st workaround is that you will not add or remove reviewers halfway, and all the required reviewers need being added at the beginning.</p>
<blockquote>
<p><strong>I tried using 2nd approach here. I get one issue here.</strong></p>
</blockquote>
<p>Sorry for my mistake, I forget that the command “<strong>ConvertTo-Json</strong>” is used in <strong>PowerShell</strong>. If you have added “<strong>Content-Type: application/json</strong>” into the header of the API, you do not need to convert the response to JSON-type again.</p>
<p>Due to the current APIs are based on the review requests, not reviewers. One user may be requested as reviewer multiple times in a PR. So if a user has been requested multiple times, the APIs will return the approvals multiple times.</p>

**Reply by manoj27730** [reply]
Can we get reviewer's email id also

**Comment by so-amuk** [comment]
<p>Thanks <a class="mention" href="/brightran">@brightran</a>  for your answer and ConvertTo-Json is working fine. Finally I am able to call API and receive data.</p>
<pre><code class="lang-auto">name: Test Github actions
on: 
    pull_request_review:
      types: [submitted]
jobs:
    triggerJenkins:
        name: Trigger Jenkins
        runs-on: ubuntu-latest
        steps:
            -   name: trigger
                shell: pwsh
                run: | 
                    $uri = "https://api.github.com/repos/xyz/project/pulls/2/reviews"
                    $headers = New-Object "System.Collections.Generic.Dictionary[[String],[String]]"
                    $headers.Add("Authorization", "token xxxxxxx")
                    $headers.Add("Content-Type", "application/json")
                    $response = Invoke-RestMethod -Uri $uri -Headers $headers -Method GET
                    echo $response | ConvertTo-Json | jq '.[] | select(.state=="APPROVED")' json
</code></pre>
<p>Now I am facing some new problem, and didn’t get any solution yet. Can you please have a look</p>
<p><strong>jq: error: APPROVED/0 is not defined at , line 1:</strong></p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/715882b7293a3e52aad7a83b5fa72841e342f8bd.png" data-download-href="/uploads/short-url/gaHApr3DhJX55myXtorFjS2iFcx.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/39334174/181090678-1d587012-ac81-4d00-936c-5302a9cbd392.png" alt="image" data-base62-sha1="gaHApr3DhJX55myXtorFjS2iFcx" width="690" height="256" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/715882b7293a3e52aad7a83b5fa72841e342f8bd_2_690x256.png, https://aws1.discourse-cdn.com/github/original/2X/7/715882b7293a3e52aad7a83b5fa72841e342f8bd.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/715882b7293a3e52aad7a83b5fa72841e342f8bd.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/715882b7293a3e52aad7a83b5fa72841e342f8bd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">896×333 23.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/so-amuk">@so-amuk</a>,<br>
Try this:</p>
<pre><code class="lang-yaml">echo $response | ConvertTo-Json | jq '.[] | select(.state=="APPROVED")'
</code></pre>
<p>Or try cURL on bash:</p>
<pre><code class="lang-yaml">      - name: List reviews on the pull request
        shell: bash
        run: |
          token=${{ secrets.GITHUB_TOKEN }}
          base64AuthInfo='$token | base64 -e'
          
          response=$(curl --request GET \
          --url https://api.github.com/repos/${{ github.repository }}/pulls/${{ github.event.pull_request.number }}/reviews \
          --header 'Authorization: $base64AuthInfo' \
          --header 'Content-Type: application/json')

          echo $response | jq '.[] | select(.state=="APPROVED")'
</code></pre>
<p>But there is a problem I found, the <strong>select</strong> function will return each of the matched objects/elements in the array as an individual object, not group them into a new array. So in this way, it’s hard to get the number of the matched objects.<br>
You may need use a loop to access the every element of the array, and check the value of <strong>state</strong> in each element and count the “<strong>APPROVED</strong>”.</p>

**Comment by so-amuk** [comment]
<p>Thanks <a class="mention" href="/brightran">@brightran</a>, really appreciated for your answer. This would help me to resolve my problem. Thanks!!</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/so-amuk">@so-amuk</a>,<br>
You’re welcome.</p>
